src/Tracker.cpp

```diff
@@ -1895,3 +1968,157 @@ void Tracker::CollectNearestPoints(KeyFrame& kf, std::set<MapPoint*>& spNearestP
+
+    double totalEntropyReduction = 0;
+    int totalEntropyPoints = 0;
+    double totalPreviousEntropy = 0;
+
+    for(unsigned i=0; i < mvCurrCamNames.size(); ++i)
+    {
+        std::string camName = mvCurrCamNames[i];
+        KeyFrame& kf = *mpTempMKF->mmpKeyFrames[camName];
+
+        for(TrackerDataPtrVector::iterator td_it = mvIterationSets[i].begin(); td_it!= mvIterationSets[i].end(); ++td_it)
+        {
+            TrackerData td = *(*td_it);
+
+            if(!td.mbFound || td.mPoint.mbBad)
+                continue;
+
+            Measurement* pMeas = new Measurement;
+            pMeas->eSource = Measurement::SRC_TRACKER;
+            pMeas->v2RootPos = td.mv2Found;
+            pMeas->nLevel = td.mnSearchLevel;
+            pMeas->bSubPix = td.mbDidSubPix;
+
+            MapPoint& point = td.mPoint;
+            kf.AddMeasurement(&point, pMeas); //add this measurement to the kf
+
+            double priorPointCovariance = point.depthCovariance;
+            double prevPointEntropy = 0;
+
+            if( !std::isfinite(priorPointCovariance) || priorPointCovariance < 1e-7) //todo (adas): don't hardcode these
+                priorPointCovariance = 1e-8;
+
+            double entropyReduction = EvaluatePoint(this, point, kf, priorPointCovariance, pMeas->nLevel, prevPointEntropy);
+
+            if(!isnan(entropyReduction) ) //if non NAN, todo (adas) need to figure out why nan, probably ill conditioned covariance
+            {
+                totalPreviousEntropy+=prevPointEntropy;
+                totalEntropyReduction +=entropyReduction;
+                totalEntropyPoints++;
+
+            }
+        }
+    }
+
+    mvKeyFrameBuffer.push_back(mpTempMKF); //store the MKF
+    //ROS_INFO("Tracker: Buffering keyframe: %ld", mvKeyFrameBuffer.size());
+    //ROS_INFO("Tracker: Total MKF Entropy Reduction is: %f", totalEntropyReduction);
+
+    score_pair mp; mp.first = totalEntropyReduction; mp.second = mvKeyFrameBuffer.size() - 1;
+    vKeyframeScores.push_back(mp);
+
+    if(totalEntropyReduction > bestBufferKeyframeScore) //if we have a new winner
+    {
+        bestBufferKeyframeScore = totalEntropyReduction;
+        bestBufferKeyframeIndex = mvKeyFrameBuffer.size() - 1;
+    }
+
+
+}
+
+void Tracker::AddNewKeyFrameFromBuffer(int bufferPosition)
+{
+    if( (int)mvKeyFrameBuffer.size() < bufferPosition) //exceeding index
+        return;
+
+    static GVars3::gvar3<int> gvnCrossCamera("CrossCamera", 1, GVars3::HIDDEN|GVars3::SILENT);
+    //note: should check to make sure index doesnt exceed buffer size
+
+    //pull out the keyframe from the buffer
+    MultiKeyFrame* mpTempMKF =  mvKeyFrameBuffer[bufferPosition];
+
+    //set it's status to NOT a bufferkeyframe (it's getting inserted into the map!
+    mpTempMKF->isBufferMKF = false;
+
+    //iterate through all the measurements in the chosen MKF and all all the measurements to the map points
+    //boost::mutex::scoped_lock lock(mMap.mMutex);
```

Comment:
line 2054: if( (int)mvKeyFrameBuffer.size() < bufferPosition) //exceeding index
        return;
Shouldn't < be <=? doesn't buffer index range from 0 to size-1, so index==size can cause overflow?